core/src/main/java/io/grpc/internal/RetryPolicy.java

```diff
@@ -0,0 +1,54 @@
+/*
+ * Copyright 2016, Google Inc. All rights reserved.
+ *
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions are
+ * met:
+ *
+ *    * Redistributions of source code must retain the above copyright
+ * notice, this list of conditions and the following disclaimer.
+ *    * Redistributions in binary form must reproduce the above
+ * copyright notice, this list of conditions and the following disclaimer
+ * in the documentation and/or other materials provided with the
+ * distribution.
+ *
+ *    * Neither the name of Google Inc. nor the names of its
+ * contributors may be used to endorse or promote products derived from
+ * this software without specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+ * "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+ * LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
+ * A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
+ * OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+ * SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+ * LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+ * DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+ * THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+package io.grpc.internal;
+
+import io.grpc.CallOptions;
+import io.grpc.MethodDescriptor;
+import io.grpc.Status;
+
+public abstract class RetryPolicy {
+  private final BackoffPolicy.Provider backoffPolicyProvider;
+
+  RetryPolicy(BackoffPolicy.Provider backoffPolicyProvider) {
+    this.backoffPolicyProvider = backoffPolicyProvider;
+  }
+
+  RetryPolicy() {
+    this(new ExponentialBackoffPolicy.Provider());
+  }
+
+  public long getNextBackoffMillis() {
+    return backoffPolicyProvider.get().nextBackoffMillis();
```

Comment:
Always calling `provider.get()` means the back-off is always reset to initial condition, so never actually growing exponentially.

You need a way to instantiate a new back-off policy per call. And then, for multiple attempts within the same call, invoke `nextBackoffMillis()`.

Also, the default settings for `ExponentialBackoffPolicy` are appropriate for connection attempts but less so for individual RPCs -- the initial delay is a whole second. (Fine for testing, but probably needs to be tuned for real use.)